File: modules/floating-bubble/android/src/main/java/expo/modules/floatingbubble/MetadataFetcher.kt
package expo.modules.floatingbubble

import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

data class ContentMetadata(
    val url: String,
    val title: String?,
    val description: String?,
    val imageUrl: String?,
    val siteName: String?,
    val platform: String,
)

object MetadataFetcher {

    private const val TIMEOUT = 8000
    private const val MAX_HEAD_BYTES = 24_000

    private val OG_TITLE = Regex("""<meta[^>]+property=["']og:title["'][^>]+content=["']([^"'<]+)["']""", RegexOption.IGNORE_CASE)
    private val OG_TITLE_ALT = Regex("""<meta[^>]+content=["']([^"'<]+)["'][^>]+property=["']og:title["']""", RegexOption.IGNORE_CASE)
    private val OG_DESC = Regex("""<meta[^>]+property=["']og:description["'][^>]+content=["']([^"'<]+)["']""", RegexOption.IGNORE_CASE)
    private val OG_DESC_ALT = Regex("""<meta[^>]+content=["']([^"'<]+)["'][^>]+property=["']og:description["']""", RegexOption.IGNORE_CASE)
    private val OG_IMAGE = Regex("""<meta[^>]+property=["']og:image["'][^>]+content=["']([^"'<]+)["']""", RegexOption.IGNORE_CASE)
    private val OG_IMAGE_ALT = Regex("""<meta[^>]+content=["']([^"'<]+)["'][^>]+property=["']og:image["']""", RegexOption.IGNORE_CASE)
    private val OG_SITE = Regex("""<meta[^>]+property=["']og:site_name["'][^>]+content=["']([^"'<]+)["']""", RegexOption.IGNORE_CASE)
    private val OG_SITE_ALT = Regex("""<meta[^>]+content=["']([^"'<]+)["'][^>]+property=["']og:site_name["']""", RegexOption.IGNORE_CASE)
    private val TITLE_TAG = Regex("""<title[^>]*>([^<]+)</title>""", RegexOption.IGNORE_CASE)

    fun fetch(url: String): ContentMetadata {
        val platform = detectPlatform(url)

        if (platform == "youtube") {
            val videoId = extractYouTubeId(url)
            val oembed = fetchYouTubeOembed(url)
            if (oembed != null) return oembed
            if (videoId != null) {
                return ContentMetadata(
                    url = url, title = null, description = null,
                    imageUrl = "https://img.youtube.com/vi/$videoId/hqdefault.jpg",
                    siteName = "YouTube", platform = platform,
                )
            }
        }

        val ogResult = fetchOgTags(url, platform)
        if (ogResult != null && (ogResult.title != null || ogResult.imageUrl != null)) return ogResult

        val microlink = fetchMicrolink(url, platform)
        if (microlink != null) return microlink

        return ContentMetadata(url = url, title = null, description = null, imageUrl = null, siteName = null, platform = platform)
    }

    private fun fetchYouTubeOembed(url: String): ContentMetadata? = try {
        val oembedUrl = "https://www.youtube.com/oembed?url=${URLEncoder.encode(url, "UTF-8")}&format=json"
        val json = getJson(oembedUrl) ?: return null
        ContentMetadata(
            url = url,
            title = json.optString("title").takeIf { it.isNotBlank() },
            description = null,
            imageUrl = json.optString("thumbnail_url").takeIf { it.isNotBlank() },
            siteName = "YouTube",
            platform = "youtube",
        )
    } catch (_: Throwable) { null }

    private fun fetchOgTags(url: String, platform: String): ContentMetadata? = try {
        val conn = (URL(url).openConnection() as HttpURLConnection).apply {
            connectTimeout = TIMEOUT
            readTimeout = TIMEOUT
            setRequestProperty("User-Agent", "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)")
            setRequestProperty("Accept", "text/html,application/xhtml+xml")
            instanceFollowRedirects = true
        }
        if (conn.responseCode !in 200..299) { conn.disconnect(); return null }

        val sb = StringBuilder()
        var bytesRead = 0
        val reader = BufferedReader(InputStreamReader(conn.inputStream, Charsets.UTF_8))
        var line: String?
        while (reader.readLine().also { line = it } != null && bytesRead < MAX_HEAD_BYTES) {
            sb.appendLine(line)
            bytesRead += line?.length ?: 0
            if (line?.contains("</head>", ignoreCase = true) == true) break
        }
        reader.close()
        conn.disconnect()

        val html = sb.toString()
        ContentMetadata(
            url = url,
            title = (OG_TITLE.find(html) ?: OG_TITLE_ALT.find(html))?.groupValues?.getOrNull(1)?.trim()
                ?: TITLE_TAG.find(html)?.groupValues?.getOrNull(1)?.trim(),
            description = (OG_DESC.find(html) ?: OG_DESC_ALT.find(html))?.groupValues?.getOrNull(1)?.trim(),
            imageUrl = (OG_IMAGE.find(html) ?: OG_IMAGE_ALT.find(html))?.groupValues?.getOrNull(1)?.trim(),
            siteName = (OG_SITE.find(html) ?: OG_SITE_ALT.find(html))?.groupValues?.getOrNull(1)?.trim(),
            platform = platform,
        )
    } catch (_: Throwable) { null }

    private fun fetchMicrolink(url: String, platform: String): ContentMetadata? = try {
        val apiUrl = "https://api.microlink.io/?url=${URLEncoder.encode(url, "UTF-8")}"
        val json = getJson(apiUrl) ?: return null
        if (json.optString("status") != "success") return null
        val d = json.optJSONObject("data") ?: return null
        val image = d.optJSONObject("image")?.optString("url")?.takeIf { it.isNotBlank() }
            ?: d.optJSONObject("logo")?.optString("url")?.takeIf { it.isNotBlank() }
        ContentMetadata(
            url = url,
            title = d.optString("title").takeIf { it.isNotBlank() },
            description = d.optString("description").takeIf { it.isNotBlank() },
            imageUrl = image,
            siteName = d.optString("publisher").takeIf { it.isNotBlank() },
            platform = platform,
        )
    } catch (_: Throwable) { null }

    private fun getJson(url: String): JSONObject? {
        val conn = (URL(url).openConnection() as HttpURLConnection).apply {
            connectTimeout = TIMEOUT
            readTimeout = TIMEOUT
            setRequestProperty("User-Agent", "Savely/1.0")
        }
        return try {
            if (conn.responseCode !in 200..299) return null
            JSONObject(conn.inputStream.bufferedReader().readText())
        } catch (_: Throwable) { null } finally { conn.disconnect() }
    }

    private fun extractYouTubeId(url: String): String? {
        val patterns = listOf(
            Regex("youtube\\.com/watch\\?v=([A-Za-z0-9_-]+)"),
            Regex("youtu\\.be/([A-Za-z0-9_-]+)"),
            Regex("youtube\\.com/shorts/([A-Za-z0-9_-]+)"),
        )
        for (p in patterns) {
            return p.find(url)?.groupValues?.getOrNull(1) ?: continue
        }
        return null
    }

    fun detectPlatform(url: String): String {
        val u = url.lowercase()
        return when {
            u.contains("youtube.com") || u.contains("youtu.be") -> "youtube"
            u.contains("instagram.com") -> "instagram"
            u.contains("tiktok.com") -> "tiktok"
            u.contains("twitter.com") || u.contains("x.com") -> "twitter"
            u.contains("reddit.com") -> "reddit"
            u.contains("pinterest.com") -> "pinterest"
            u.contains("linkedin.com") -> "linkedin"
            u.contains("facebook.com") || u.contains("fb.com") -> "facebook"
            u.contains("threads.net") -> "threads"
            u.contains("spotify.com") -> "spotify"
            u.contains("maps.google") || u.contains("google.com/maps") -> "maps"
            else -> "link"
        }
    }
}

File: modules/savely-share-extension/ios/ShareExtension/MetadataFetcher.swift
import Foundation

struct ContentMetadata {
  let url: String
  let title: String?
  let description: String?
  let imageUrl: String?
  let siteName: String?
  let platform: String
}

enum MetadataFetcher {
  private static let timeout: TimeInterval = 8
  private static let maxHeadBytes = 24_000

  static func detectPlatform(_ url: String) -> String {
    let u = url.lowercased()
    if u.contains("youtube.com") || u.contains("youtu.be") { return "youtube" }
    if u.contains("instagram.com") { return "instagram" }
    if u.contains("tiktok.com") { return "tiktok" }
    if u.contains("twitter.com") || u.contains("x.com") { return "twitter" }
    if u.contains("reddit.com") { return "reddit" }
    if u.contains("pinterest.com") { return "pinterest" }
    if u.contains("linkedin.com") { return "linkedin" }
    if u.contains("facebook.com") || u.contains("fb.com") { return "facebook" }
    if u.contains("threads.net") { return "threads" }
    if u.contains("spotify.com") { return "spotify" }
    if u.contains("maps.google") || u.contains("google.com/maps") { return "maps" }
    return "link"
  }

  static func fetch(url: String) async -> ContentMetadata {
    let platform = detectPlatform(url)

    if platform == "youtube" {
      if let oembed = await fetchYouTubeOembed(url: url) { return oembed }
      if let id = extractYouTubeId(url) {
        return ContentMetadata(
          url: url, title: nil, description: nil,
          imageUrl: "https://img.youtube.com/vi/\(id)/hqdefault.jpg",
          siteName: "YouTube", platform: platform
        )
      }
    }

    let blockedByAuth: Set<String> = ["instagram", "twitter", "tiktok", "linkedin", "facebook", "threads"]
    if blockedByAuth.contains(platform) {
      if let m = await fetchMicrolink(url: url, platform: platform) { return m }
      return ContentMetadata(url: url, title: nil, description: nil, imageUrl: nil, siteName: nil, platform: platform)
    }

    let og = await fetchOgTags(url: url, platform: platform)
    if let og = og, og.imageUrl != nil { return og }

    if let m = await fetchMicrolink(url: url, platform: platform) { return m }
    return og ?? ContentMetadata(url: url, title: nil, description: nil, imageUrl: nil, siteName: nil, platform: platform)
  }

  // MARK: - YouTube oEmbed

  private static func fetchYouTubeOembed(url: String) async -> ContentMetadata? {
    guard let encoded = url.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
          let oembedUrl = URL(string: "https://www.youtube.com/oembed?url=\(encoded)&format=json") else { return nil }
    guard let json = await getJson(url: oembedUrl) else { return nil }
    let title = (json["title"] as? String).flatMap { $0.isEmpty ? nil : $0 }
    let image = (json["thumbnail_url"] as? String).flatMap { $0.isEmpty ? nil : $0 }
    return ContentMetadata(url: url, title: title, description: nil, imageUrl: image, siteName: "YouTube", platform: "youtube")
  }

  // MARK: - OG tags

  private static func fetchOgTags(url: String, platform: String) async -> ContentMetadata? {
    guard let pageUrl = URL(string: url) else { return nil }
    var req = URLRequest(url: pageUrl, timeoutInterval: timeout)
    req.setValue("Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)", forHTTPHeaderField: "User-Agent")
    req.setValue("text/html,application/xhtml+xml", forHTTPHeaderField: "Accept")
    do {
      let (data, response) = try await URLSession.shared.data(for: req)
      guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else { return nil }
      let limited = data.prefix(maxHeadBytes)
      guard let html = String(data: Data(limited), encoding: .utf8) else { return nil }
      let title = firstMatch(html: html, primary: ogRegex("og:title"), alt: ogRegexAlt("og:title")) ?? titleTag(html: html)
      let desc = firstMatch(html: html, primary: ogRegex("og:description"), alt: ogRegexAlt("og:description"))
      let image = firstMatch(html: html, primary: ogRegex("og:image"), alt: ogRegexAlt("og:image"))
      let site = firstMatch(html: html, primary: ogRegex("og:site_name"), alt: ogRegexAlt("og:site_name"))
      return ContentMetadata(url: url, title: title, description: desc, imageUrl: image, siteName: site, platform: platform)
    } catch {
      return nil
    }
  }

  // MARK: - Microlink fallback

  private static func fetchMicrolink(url: String, platform: String) async -> ContentMetadata? {
    guard let encoded = url.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
          let apiUrl = URL(string: "https://api.microlink.io/?url=\(encoded)") else { return nil }
    guard let json = await getJson(url: apiUrl) else { return nil }
    guard (json["status"] as? String) == "success",
          let data = json["data"] as? [String: Any] else { return nil }
    let image = (data["image"] as? [String: Any])?["url"] as? String
      ?? (data["logo"] as? [String: Any])?["url"] as? String
    return ContentMetadata(
      url: url,
      title: (data["title"] as? String).flatMap { $0.isEmpty ? nil : $0 },
      description: (data["description"] as? String).flatMap { $0.isEmpty ? nil : $0 },
      imageUrl: image?.isEmpty == false ? image : nil,
      siteName: (data["publisher"] as? String).flatMap { $0.isEmpty ? nil : $0 },
      platform: platform
    )
  }

  // MARK: - Helpers

  private static func getJson(url: URL) async -> [String: Any]? {
    var req = URLRequest(url: url, timeoutInterval: timeout)
    req.setValue("Savely/1.0", forHTTPHeaderField: "User-Agent")
    do {
      let (data, response) = try await URLSession.shared.data(for: req)
      guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else { return nil }
      return try JSONSerialization.jsonObject(with: data) as? [String: Any]
    } catch {
      return nil
    }
  }

  private static func ogRegex(_ property: String) -> String {
    return "<meta[^>]+property=[\"']\(property)[\"'][^>]+content=[\"']([^\"'<]+)[\"']"
  }

  private static func ogRegexAlt(_ property: String) -> String {
    return "<meta[^>]+content=[\"']([^\"'<]+)[\"'][^>]+property=[\"']\(property)[\"']"
  }

  private static func firstMatch(html: String, primary: String, alt: String) -> String? {
    if let m = matchFirst(pattern: primary, in: html) { return m.trimmingCharacters(in: .whitespacesAndNewlines) }
    if let m = matchFirst(pattern: alt, in: html) { return m.trimmingCharacters(in: .whitespacesAndNewlines) }
    return nil
  }

  private static func titleTag(html: String) -> String? {
    matchFirst(pattern: "<title[^>]*>([^<]+)</title>", in: html)?.trimmingCharacters(in: .whitespacesAndNewlines)
  }

  private static func matchFirst(pattern: String, in text: String) -> String? {
    guard let re = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else { return nil }
    let range = NSRange(text.startIndex..<text.endIndex, in: text)
    guard let m = re.firstMatch(in: text, options: [], range: range), m.numberOfRanges >= 2 else { return nil }
    guard let r = Range(m.range(at: 1), in: text) else { return nil }
    return String(text[r])
  }

  private static func extractYouTubeId(_ url: String) -> String? {
    let patterns = [
      "youtube\\.com/watch\\?v=([A-Za-z0-9_-]+)",
      "youtu\\.be/([A-Za-z0-9_-]+)",
      "youtube\\.com/shorts/([A-Za-z0-9_-]+)",
    ]
    for p in patterns {
      if let id = matchFirst(pattern: p, in: url) { return id }
    }
    return nil
  }
}
